File: my-solutions/0042-trapping-rain-water/solution.py
class Solution:
    def trap(self, height: List[int]) -> int:
        maxR = [float(-inf)] * len(height)
        maxL = [float(-inf)] * len(height)
        currMaxL, currMaxR = float(-inf),  float(-inf)
        water = 0
        for i in range(len(height)):
            currMaxL = max(currMaxL, height[i])
            maxL[i] = currMaxL

            currMaxR = max(currMaxR, height[len(height) - 1 - i])
            maxR[len(height) - 1 - i] = currMaxR

        for i in range(1, len(height) - 1):
            if maxR[i] > height[i] and maxL[i] > height[i]:
                water += min(maxR[i], maxL[i]) - height[i]

        return water

File: my-solutions/0042-trapping-rain-water/solution.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        vector<int> maxLeft(n), maxRight(n);
        int res = 0;

        int currMax = 0;
        for(int i = 0; i < n; ++i) {
            maxLeft[i] = currMax;
            currMax = max(currMax, height[i]);
        }

        currMax = 0;
        for(int i = n - 1; i > -1; --i) {
            maxRight[i] = currMax;
            currMax = max(currMax, height[i]);
        }

        for(int i = 0; i < n; ++i) {
            int maxWater = min(maxLeft[i], maxRight[i]);
            res += maxWater > height[i] ? maxWater - height[i] : 0;
        }

        return res;
    }
};
